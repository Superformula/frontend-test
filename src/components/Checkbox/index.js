import React from 'react';
import PropTypes from 'prop-types';
import { Wrapper } from './Wrapper';
import { Input } from './Input';
import { Label } from './Label';

export const Checkbox = ({ id, onChange, children }) => (
  <Wrapper>
    <Input {...{ onChange, id }} />
    <Label htmlFor={id}>{children}</Label>
  </Wrapper>
)

Checkbox.displayName = 'Checkbox';

Checkbox.propTypes = {
  label: PropTypes.string,
  id: PropTypes.string.isRequired,
};
