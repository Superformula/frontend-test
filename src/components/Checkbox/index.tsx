import React from "react";
import styled from "@emotion/styled";

// if we need super extra flexibilty we could
// accept a prop called inputProps of the type
// React.InputHTMLAttributes<HTMLInputElement>
// and so we can give full control of the underlying
// component but Im going to take a more minimalistic
// approach of only accepting the needed props
export interface IProps {
  checked: boolean;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => any;
  /* specify the label */
  children: React.ReactNode;
}

// TODO improve the internal layout so that it is
// probably relative and avoids using so many hardcoded values.
// The approach I am taking is similar to that which Ant design
// uses and a lot of resources in the internet point out as the
// main way of achieving this, but with more time I would attempt
// to rely more on relative values and avoid harcoded values.
export default function Checkbox(props: IProps) {
  return (
    <LabelStyled>
      <CheckboxStyled
        type="checkbox"
        checked={props.checked}
        onChange={props.onChange}
      />
      <FakeCheckbox />
      <InputLabel>{props.children}</InputLabel>
    </LabelStyled>
  );
}

export const CIRCLE_SIZE_PX = 16;

export const InputLabel = styled.span`
  color: ${({ theme }) => theme.colors.primary};
  font-size: ${({ theme }) => theme.fontSize.x600};
`;

export const FakeCheckbox = styled.span`
  position: absolute;
  left: 0px;
  // TODO I don't live this hardcoded pixels
  top: calc(50% - ${CIRCLE_SIZE_PX / 2}px);
  height: ${CIRCLE_SIZE_PX}px;
  width: ${CIRCLE_SIZE_PX}px;
  background-color: white;
  border: 1px solid ${({ theme }) => theme.colors.primary};
  border-radius: 50%;

  /* Create the checkmark/indicator (hidden when not checked) */
  &:after {
    content: "";
    position: absolute;
    display: none;
  }
`;

export const LabelStyled = styled.label`
  position: relative;
  // 8 is the padding between the circle
  // and the label
  // TODO un hardcode this
  padding-left: ${CIRCLE_SIZE_PX + 8}px;
  padding-top: 8px;
  padding-bottom: 8px;
  border-bottom: 1px solid ${({ theme }) => theme.colors.grey200};
  cursor: pointer;
  font-size: ${({ theme }) => theme.fontSize.x600};
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  /* On mouse-over, add a grey background color */
  &:hover input ~ ${FakeCheckbox} {
    background-color: #ccc;
  }

  /* When the checkbox is checked, add a blue background */
  & input:checked ~ ${FakeCheckbox} {
  }

  /* Show the checkmark when checked */
  & input:checked ~ ${FakeCheckbox}:after {
    display: block;
    background-color: ${({ theme }) => theme.colors.primary};
  }

  /* Style the checkmark/indicator */
  & ${FakeCheckbox}:after {
    // TODO more magic constants!
    left: 1px;
    top: 1px;
    width: 8px;
    height: 8px;
    border: solid white;
    border-radius: 50%;
  }
`;

// Hide the real checkbox and only rely
// on it's internal state
export const CheckboxStyled = styled.input`
  position: absolute;
  opacity: 0;
  cursor: pointer;
  height: 0;
  width: 0;
`;
