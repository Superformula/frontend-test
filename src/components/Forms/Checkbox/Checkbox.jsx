/* eslint-disable jsx-a11y/label-has-for */
// TODO: ^ we shouldn't have to do this.
// inspiration for this component was drawn from:
// https://medium.com/@colebemis/building-a-checkbox-component-with-react-and-styled-components-8d3aa1d826dd
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { borders, borderColor } from 'styled-system';

import { Label } from '../..';

const CheckboxContainer = styled.div`
  display: inline-flex;
  align-items: center;
`;

const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  // Source: https://polished.js.org/docs/#hidevisually
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;

const StyledCheckbox = styled.div`
  display: inline-block;
  width: 14px;
  height: 14px;
  border-radius: 50%;
  ${borders}
  ${borderColor}
  position: relative;
`;

const Icon = styled.svg`
  position: absolute;
  > circle {
     transition: all ease .25s;
     stroke: ${props => props.theme.colors.secondary};
     stroke-width: ${props => props.checked ? '4px' : '0'};
     fill: ${props => props.checked ? props.theme.colors.primary : props.theme.colors.secondary};
  }
`;

const Checkbox = ({
  checked,
  label,
  name,
  ...props
}) => (
  <CheckboxContainer>
    <Label
      htmlFor={name}
      borderBottom="1px solid"
      borderColor="black21"
      pb={2}
    >
      <HiddenCheckbox checked={checked} name={name} {...props} />
      <StyledCheckbox
        checked={checked}
        border="1px solid"
        borderColor="black21"
      >
        <Icon viewBox="0 0 24 24" checked={checked}>
          <circle cx="12" cy="12" r="10" />
        </Icon>
      </StyledCheckbox>
      <span>{label}</span>
    </Label>
  </CheckboxContainer>
);

Checkbox.defaultProps = {
  label: undefined,
  checked: false,
  isDisabled: false,
  className: undefined,
};

Checkbox.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  checked: PropTypes.bool,
  isDisabled: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  className: PropTypes.string,
};

export default Checkbox;
