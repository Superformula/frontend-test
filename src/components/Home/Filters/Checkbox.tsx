import * as React from "react";

import styles from "./Filters.css";

interface CheckboxProps {
  isChecked: boolean;
  handleCheck: () => void;
  label: string;
}

const Checkbox: React.FC<CheckboxProps> = ({ isChecked, handleCheck, label }) => {
  return (
    <div onClick={handleCheck} className={styles.filter}>
      {isChecked ? <FullIcon /> : <EmptyIcon />}
      <label>{label}</label>
    </div>
  );
};

export default Checkbox;

// Icons

const EmptyIcon: React.FC = () => {
  return (
    <svg
      className={styles.checkboxIcon}
      xmlns="http://www.w3.org/2000/svg"
      enableBackground="new 0 0 426.667 426.667"
      version="1.1"
      viewBox="0 0 426.67 426.67"
      xmlSpace="preserve"
    >
      <path d="M213.333,0C95.467,0,0,95.467,0,213.333s95.467,213.333,213.333,213.333S426.667,331.2,426.667,213.333S331.2,0,213.333,0    z M213.333,384c-94.293,0-170.667-76.373-170.667-170.667S119.04,42.667,213.333,42.667S384,119.04,384,213.333    S307.627,384,213.333,384z" />
    </svg>
  );
};

const FullIcon: React.FC = () => {
  return (
    <svg
      className={styles.checkboxIcon}
      enableBackground="new 0 0 426.667 426.667"
      version="1.1"
      viewBox="0 0 426.667 426.667"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path d="m213.33 106.67c-58.88 0-106.67 47.787-106.67 106.67s47.787 106.67 106.67 106.67 106.67-47.787 106.67-106.67-47.787-106.67-106.67-106.67z" />
      <path d="M213.333,0C95.467,0,0,95.467,0,213.333s95.467,213.333,213.333,213.333S426.667,331.2,426.667,213.333     S331.2,0,213.333,0z M213.333,384c-94.293,0-170.667-76.373-170.667-170.667S119.04,42.667,213.333,42.667     S384,119.04,384,213.333S307.627,384,213.333,384z" />
    </svg>
  );
};
