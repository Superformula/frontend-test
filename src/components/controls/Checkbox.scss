@import '../../scss/variables';

.checkbox-container {
    input[type='checkbox'] {
        display: none;
    }
    .circle-checkbox {
        float: left;
        width: 16px;
        height: 16px;
        background: $color-white;
        border-radius: 100%;
        border: 1px solid $color-light-grey;
        position: relative;
        margin-top: 2px;
    }
    .circle-checkbox label {
        display: block;
        width: 8px;
        height: 8px;
        border-radius: 100px;

        transition: all 0.2s ease;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);

        background: $color-white;
        pointer-events: none;
    }
    .circle-checkbox input[type='checkbox']:checked + label {
        background: $color-primary;
    }
    .label-text {
        font-size: $font-size-label;
        color: $color-primary;
        display: inline-block;
        margin-left: 8px;
    }
}
