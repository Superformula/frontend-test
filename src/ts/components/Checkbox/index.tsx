// Imports
// -------

// Libraries
import React from 'react';
// Images
import checkedCheckbox from '../../../img/checkedCheckbox.svg';
import emptyCheckbox from '../../../img/emptyCheckbox.svg';
import filledCheckbox from '../../../img/filledCheckbox.svg';
// Styles
import style from './style.module.css';

// Internal
// --------

export interface CheckboxProps {
    checked?: boolean
    children: string,
    role?: 'checkbox' | 'radiobutton',
    onClick?(): any
}

const Checkbox: React.FC<CheckboxProps> = ({checked, role = 'checkbox', onClick, children}) => {
    //
    // Render

    const imgSrc = checked ? (role === 'checkbox' ? checkedCheckbox : filledCheckbox) : emptyCheckbox;

    return (
        <label className={ style.checkbox } onClick={ onClick }>
            <img src={ imgSrc } alt={ `${children} ${checked ? 'checked' : ''}` } className={ style.icon }/>
            {/*<input type="checkbox"/>*/}
            { children }
        </label>
    );
}

// Exports
// -------

export default Checkbox;
