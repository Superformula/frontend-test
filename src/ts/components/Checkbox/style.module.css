.checkbox {
    display: inline-flex;
    align-items: center;
    color: #002B56;
    cursor: pointer;
    padding: 8px 0;
}

.icon {
    margin-right: 8px;
}
