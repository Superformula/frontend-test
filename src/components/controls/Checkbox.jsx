import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './Checkbox.scss';
import { controlContainer } from './Controls.module.scss';
const Checkbox = ({ label, checked, onChange }) => {
    return (
        <div
            className={classNames(controlContainer, 'checkbox-container')}
            onClick={() => {
                onChange(!checked);
            }}
        >
            <div className="circle-checkbox">
                <input type="checkbox" checked={checked} id="checkboxInput" readOnly />
                <label htmlFor="checkboxInput" />
            </div>
            <div className="label-text">{label}</div>
        </div>
    );
};

Checkbox.propTypes = {
    label: PropTypes.string,
    checked: PropTypes.bool,
    onChange: PropTypes.func
};
export default Checkbox;
