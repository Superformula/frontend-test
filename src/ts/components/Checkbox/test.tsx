// Imports
// -------

// Libraries
import React from 'react';
import renderer from 'react-test-renderer';
// Components
import Checkbox from './index';

// Mocks
// -----

jest.mock('../../../img/checkedCheckbox.svg', () => 'checkedCheckbox.svg');
jest.mock('../../../img/emptyCheckbox.svg', () => 'emptyCheckbox.svg');
jest.mock('../../../img/filledCheckbox.svg', () => 'filledCheckbox.svg');

// Internal
// --------

const children = "Test";

const getComponent = (props = {}) => renderer.create(
    <Checkbox {...props} >
        { children }
    </Checkbox>,
);

// Tests
// -----

describe('src/ts/components/Checkbox', () => {
    let component;

    describe('Rendering', () => {
        it('renders default correctly', () => {
            component = getComponent();
            expect(component).toMatchSnapshot();
        });
    });
});
