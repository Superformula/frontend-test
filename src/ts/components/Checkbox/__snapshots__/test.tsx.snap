// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/ts/components/Checkbox Rendering renders default correctly 1`] = `
<label
  className="checkbox"
>
  <img
    alt="Test "
    className="icon"
    src="emptyCheckbox.svg"
  />
  Test
</label>
`;
